components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/inbound/IdentityRequest.java

```diff
@@ -283,6 +285,31 @@ public IdentityRequestBuilder addParameters(Map<String, String[]> parameters) {
             return this;
         }
 
+        public IdentityRequestBuilder setAttributes(Map<String, Object> attributes) {
+            this.attributes = new HashMap(attributes);;
+            return this;
+        }
+
+        public IdentityRequestBuilder addAttribute(String name, Object value) {
+            if (this.attributes.containsKey(name)) {
+                throw FrameworkRuntimeException.error("Attributes map trying to override existing " +
+                                                      "key " + name);
+            }
+            this.attributes.put(name, value);
+            return this;
+        }
+
+        public IdentityRequestBuilder addAttributes(Map<String, Object> attributes) {
+            for (Map.Entry<String, Object> attribute : attributes.entrySet()) {
+                if (this.attributes.containsKey(attribute.getKey())) {
```

Comment:
shall we replace 304-308 with `addAttribute(attribute.getKey(), attribute.getValue())` ?